### ISSUE TITLE
Improve ResNet training regime

### ISSUE THREAD

**Issue by philswatton** [alan-turing-institute/arc-model-similarities-phase-2#26]
Since model similarity phase 1, some work has been done on how best to train ResNet. We should check to ensure that insofar as is appropriate (i.e. given the modifications we've made for CIFAR-10), we are training ResNet in the best possible way.

In particular, start with this article: <https://arxiv.org/abs/2110.00476>

**Comment by philswatton** [comment]
Linking the paper on relating learning rates to batch sizes here, as it should be part of the PR that implements this issue https://arxiv.org/pdf/2006.09092.pdf

**Comment by philswatton** [comment]
Broad findings in terms of learning rate/batch size:
- for Stochastic Gradient Descent, several papers show that the learning rate scales linearly with batch size (https://arxiv.org/pdf/2006.09092.pdf from above, https://arxiv.org/pdf/1711.00489.pdf, https://arxiv.org/pdf/1706.02677.pdf)
  - The papers noting this are:
    - Learning Rates as a Function of Batch Size: A Random Matrix Theory Approach to Neural Network Training (https://arxiv.org/pdf/2006.09092.pdf) (paper cited in the original issue)
    - Don't Decay the Learning Rate, Increase the Batch Size (https://arxiv.org/pdf/1711.00489.pdf)
    - Accurate, Large Minibatch SGD: Training Imagenet in one hour (https://arxiv.org/pdf/1706.02677.pdf)
  - None of these papers give specifics on a universal baseline number though (e.g. lr = x for batch size b)
  - The third paper however does suggest that for their work, for a batch size of 256 the learning rate should be 0.1. Applying a linear scaling rule, this implies that for a batch size of 32 the learning rate should be  0.0125
  
timm paper (https://arxiv.org/pdf/2110.00476.pdf) (listed above) findings:
- this paper is broadly about ResNet50
- it describes three procedures, A1, A2, A3. A1 takes the longest (600 epochs) but achieves best performance, A3 takes the shortest (100 epochs, 15hrs on their machine) and does well (but not quite as well as A1)
- part of the procedure uses data augmentations, which we should ignore as this is part of our experiments in terms of dataset similarity
- In terms of regularization, they express a general rule of thumb which is deeper networks/longer training -> more regularization. Out current training regime is currently only 10 epochs though, and I'm unsure whether we should therefore be doing more regulization, but I'm inclined to say not
- They found that it was difficult to reach convergence with the SGD optimizer. Instead, with large batch size and BCE loss, they found LAMB to be most effective
  - again, given this is predicated on the use of large batch sizes (e.g. 2048), it isn't clear to me how relevant this is to our case

Phase 1 Approach (note the idea here was to try different models/combinations)
- Stochastic Gradient Descent with Momentum 0.9
- Learning Rate Varied according to One-Cycle Annealing w/ maximum learning rate of 0.1 (https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.OneCycleLR.html)
  - This part is what I think allows for fast training of models, which we may wish to prioritise over SOTA performance given the large number of them we need to train
  - Paper: https://arxiv.org/pdf/1708.07120.pdf
- Maximum of 100 epochs
- Considered 3 levels of weight decay (0, 0.001, 0.0003)
- Trained with and without early stopping based on validation loss with patience of 7 epochs

**Comment by philswatton** [comment]
Closed without many drastic changes to the original training regime in the end, following review of some alternatives. See PR #46
